pkg/destroy/aws/aws.go

```diff
@@ -641,35 +646,110 @@ func deleteEC2Instance(ec2Client *ec2.EC2, iamClient *iam.IAM, id string, logger
 
 	for _, reservation := range response.Reservations {
 		for _, instance := range reservation.Instances {
-			// Skip 'terminated' instances since they take a while to really get cleaned up
-			if *instance.State.Name == "terminated" {
-				continue
-			}
-			if instance.IamInstanceProfile != nil {
-				parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
-				if err != nil {
-					return errors.Wrap(err, "parse ARN for IAM instance profile")
-				}
-
-				err = deleteIAMInstanceProfile(iamClient, parsed, logger)
-				if err != nil {
-					return errors.Wrapf(err, "deleting %s", parsed.String())
-				}
-			}
-
-			_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
-				InstanceIds: []*string{instance.InstanceId},
-			})
+			err = deleteEC2InstanceByInstance(ec2Client, iamClient, instance, logger)
 			if err != nil {
 				return err
 			}
+		}
+	}
+	return nil
+}
+
+func deleteEC2InstanceByInstance(ec2Client *ec2.EC2, iamClient *iam.IAM, instance *ec2.Instance, logger logrus.FieldLogger) error {
+	// Skip 'terminated' instances since they take a while to really get cleaned up
+	if instance.State == nil || *instance.State.Name == "terminated" {
+		return nil
+	}
 
-			logger.Info("Deleted")
+	if instance.IamInstanceProfile != nil {
+		parsed, err := arn.Parse(*instance.IamInstanceProfile.Arn)
+		if err != nil {
+			return errors.Wrap(err, "parse ARN for IAM instance profile")
+		}
+
+		err = deleteIAMInstanceProfile(iamClient, parsed, logger)
+		if err != nil {
+			return errors.Wrapf(err, "deleting %s", parsed.String())
 		}
 	}
+
+	if *instance.State.Name == "shutting-down" {
+		return nil
+	}
+
+	_, err := ec2Client.TerminateInstances(&ec2.TerminateInstancesInput{
+		InstanceIds: []*string{instance.InstanceId},
+	})
+	if err != nil {
+		return err
+	}
+
+	logger.Info("Terminating")
 	return nil
 }
 
+// deleteEC2InstancesByTags loops until there all instances which
+// match the given tags are terminated.
+func deleteEC2InstancesByTags(ec2Client *ec2.EC2, iamClient *iam.IAM, filters []Filter, logger logrus.FieldLogger) error {
```

Comment:
resourcegrouptaggingspi loop already supports resource type filters.. I think it would be better if we use less number of loops that do `or filtering` themselves.. let changes of error.

Also this doesn't go into `deleted` so the resourcegrouptaggingspi loop might try to send deletes good these again..

Reply:
> resourcegrouptaggingspi loop already supports resource type filters..

Sharing code is great, but running `DescribeInstancesPages` in `deleteEC2InstancesByTags` saves us from calling `DescribeInstances` in `deleteEC2Instance` for each returned instance.  Did you have other ideas for restructuring `deleteEC2InstancesByTags` to leverage existing code?

> Also this doesn't go into deleted so the resourcegrouptaggingspi loop might try to send deletes good these again.

I've added 53a8954e36 to this PR to address that.  But as I point out in that commit message, I think all that does is save us a `DescribeInstances` call (still useful).  And if we want it for `deleteEC2InstancesByTags`, we might also want it for `deleteEC2NATGatewaysByVPC` and similar.  It's probably a per-pathway call about whether saving a delete-gating describe call is worth the cost of plumbing through the deleted-ARNs data.